src/workerd/api/data-url.c++

```diff
@@ -0,0 +1,104 @@
+#include "data-url.h"
+#include <kj/encoding.h>
+
+namespace workerd::api {
+
+kj::Maybe<DataUrl> DataUrl::tryParse(kj::StringPtr url) {
+  KJ_IF_SOME(url, jsg::Url::tryParse(url)) {
+    return from(url);
+  }
+  return kj::none;
+}
+
+kj::Maybe<DataUrl> DataUrl::from(const jsg::Url& url) {
+  if (url.getProtocol() != "data:"_kj) return kj::none;
+  auto clone = url.clone(jsg::Url::EquivalenceOption::IGNORE_FRAGMENTS);
+
+  // Remove the "data:" prefix.
+  auto href = clone.getHref().slice(5);
+
+  // We scan for the first comma, which separates the MIME type from the data.
+  // Per the fetch spec, it doesn't matter if the comma is within a quoted
+  // string value in the MIME type... which is fun.
+
+  static const auto isAsciiWhitespace = [](auto c) {
+    return c == 0x09 /* tab */ ||
+          c == 0x0a /* lf  */ ||
+          c == 0x0c /* ff  */ ||
+          c == 0x0d /* cr  */ ||
+          c == 0x20 /* sp  */;
+  };
+
+  static const auto trim = [](auto label) {
+    size_t start = 0;
+    auto end = label.size();
+    while (start < end && isAsciiWhitespace(label[start])) { start++; }
+    while (end > start && isAsciiWhitespace(label[end - 1])) { end--; }
+    return label.slice(start, end).asChars();
+  };
+
+  static const auto strip = [](auto label) {
+    auto result = kj::heapArray<kj::byte>(label.size());
+    size_t len = 0;
+    for (auto c : label) {
+      if (!isAsciiWhitespace(c)) {
+        result[len++] = c;
+      }
+    }
+    return result.first(len).attach(kj::mv(result));
+  };
+
+  static const auto isBase64 = [](kj::ArrayPtr<const char> label) -> bool {
+    KJ_IF_SOME(pos, label.findLast(';')) {
+      auto res = trim(label.slice(pos + 1));
+      return res.size() == 6 &&
+          (res[0] | 0x20) == 'b' &&
+          (res[1] | 0x20) == 'a' &&
+          (res[2] | 0x20) == 's' &&
+          (res[3] | 0x20) == 'e' &&
+          (res[4] == '6') &&
+          (res[5] == '4');
+    }
+    return false;
+  };
+
+  static const auto create = [](auto input, auto decoded) {
+    KJ_IF_SOME(parsed, MimeType::tryParse(input)) {
+      return DataUrl(kj::mv(parsed), kj::mv(decoded));
+    } else {
+      return DataUrl(MimeType::PLAINTEXT_ASCII.clone(), kj::mv(decoded));
+    }
+  };
+
+  KJ_IF_SOME(pos, href.findFirst(',')) {
+    auto unparsed = href.first(pos);
+    auto data = href.slice(pos + 1);
+
+    // We need to trim leading and trailing whitespace from the mimetype
+    unparsed = trim(unparsed);
```

Comment:
Nit: Maybe it's better to name these variables the same as what they're in the fetch standard, just to make it easy to ensure that we're following the spec to a T.

Reply:
I tend to prefer not, to be honest. I'm not always a big fan of how the specs name these things and prefer the names here.